### ISSUE TITLE
Add Time value type in System

### ISSUE THREAD

**Issue by Gimly** [dotnet/corefx#28935]
# Rationale
Today, we have a `DateTime` struct in System that represents a specific date and time, which supports time zones. We also have a `TimeSpan` object that represents a duration (for example the time between two `DateTime`). I feel that we are missing a simple `Time` object that represents a time of day, without it being linked to a particular day.

There are a lot of reasons for representing time without dates. Alarms of an alarm clock, work shifts in a company, etc. T-Sql has a [time](https://docs.microsoft.com/en-us/sql/t-sql/data-types/time-transact-sql) type that cannot today be mapped to a type in .Net.

`DateTime.TimeOfDay` today returns a  `TimeSpan`, but it feels out of place, `TimeSpan` represents a duration and not a specific time of day.

## Implementation

Implementation would be pretty simple, it would have properties Hours, Minutes, Seconds, Milliseconds, Ticks and a few comparison / transformation methods.

`DateTime.TimeOfDay` would return this new `Time` struct instead of the current `TimeSpan`.

**Comment by khellang** [comment]
This is already implemented in [corefxlab](https://github.com/dotnet/corefxlab/tree/8309ca4e7a7ba30829e355631349b20f0ca9da2e/src/System.Time). Also, see https://github.com/dotnet/corefx/issues/2116 and https://github.com/dotnet/corefxlab/issues/1740 it'll probably make its way into corefx Very Soon™ 😄

**Comment by jnm2** [comment]
> `DateTime.TimeOfDay` would return this new `Time` struct instead of the current `TimeSpan`.

I totally wish this was possible, but that would make people's existing code crash at runtime or break on recompile:

```cs
TimeSpan timeOfDay = CalculateDateTime().TimeOfDay;
// Use timeOfDay
```

**Comment by khellang** [comment]
The current proposed implementation of `System.Time` has an [explicit cast from `TimeSpan`](https://github.com/dotnet/corefxlab/blob/8309ca4e7a7ba30829e355631349b20f0ca9da2e/src/System.Time/System/Time.cs#L1586-L1597), [which calls this out](https://github.com/dotnet/corefxlab/blob/8309ca4e7a7ba30829e355631349b20f0ca9da2e/src/System.Time/System/Time.cs#L1575-L1581)

**Comment by Gimly** [comment]
Great! Didn't find #2116 when I did a search, clear that I wasn't the first to think about that issue :smile:.

So... should I close this issue?

**Comment by AlexRadch** [comment]
I do not see any advantages on TimeSpan. It seems like to add FourBits type. Also some days can have additional seconds https://en.wikipedia.org/wiki/Leap_second .
Can TimeOfDay return more then 24 hours or not?

**Comment by khellang** [comment]
I don't think so. None of the BCL date and time types supports leap seconds, AFAIK.

**Comment by jnm2** [comment]
Noda Time doesn't support leap seconds either.

**Comment by Clockwork-Muse** [comment]
Also, for the most part, at the level we're dealing with, we're only concentrating on "human" time, where, while leap seconds are part of (some standards of) official civil time, most people don't know or care about them.

That's if the host system even gives you a way to detect them in the first place.  On Windows, it essentially says, "hey, the value I got from NTP says I'm a second off now, for some unknown reason, let me do what I normally do when I'm a second off, and gradually align my clock".  I'd imagine some Linux distributions do something similar.

**Comment by GSPP** [comment]
A `Date` data type would be handy as well.

**Comment by ilmax** [comment]
Can anyone from MS shed a bit of light about what's happening with System.Date and System.Time? They are in corefxlab since a long tim, but I wasn't able to find any news about promoting them to the BCL.

Are they gonna make it? It would be nice to have a type that represent a date as a first class citizen in the BCL
